src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -670,18 +689,75 @@ private Compilation CreateEmptyCompilation()
                 return finalState.HasSuccessfullyLoaded;
             }
 
-            public ICompilationTracker FreezePartialState(CancellationToken cancellationToken)
+            public ICompilationTracker WithCreationPolicy(bool create, bool forceRegeneration, CancellationToken cancellationToken)
+            {
+                return create
+                    ? WithCreateCreationPolicy(forceRegeneration)
+                    : WithDoNotCreateCreationPolicy(forceRegeneration, cancellationToken);
+            }
+
+            public ICompilationTracker WithCreateCreationPolicy(bool forceRegeneration)
+            {
+                var state = this.ReadState();
+
+                var desiredCreationPolicy = CreationPolicy.Create;
+
+                // If we've computed no state yet there's nothing to do.  This state will automatically transition 
+                // to an InProgressState with a creation policy of 'Create' anyways.
+                if (state is null)
+                    return this;
+
+                // If we're already in the state where we are running generators and skeletons (and we're not forcing
+                // regeneration) we don't need to do anything and can just return ourselves. The next request to create
+                // the compilation will do so fully.
+                if (state.CreationPolicy == desiredCreationPolicy && !forceRegeneration)
+                    return this;
+
+                // If we're forcing regeneration then we have to drop whatever driver we have so that we'll start from
+                // scratch next time around.
+                var desiredGeneratorInfo = forceRegeneration ? state.GeneratorInfo with { Driver = null } : state.GeneratorInfo;
```

Comment:
what does dropping the driver actually mean?  is it 'reloading' the source generator (if the sg generator changes), or is it that the sg documents are dropped entirely and forces generators to run no matter what the next time the compilation is requested?

Reply:
It's the latter.  Note in the code below that we trsnsition to the InProgressState.  The next time we ask for the compilation for this tracker, we will transition that InProgressState to a FinalState, which involves getting the generated docs and adding to the primordial-compilation.

Doing that calls into the driver to do the work, and the driver is what does things like incremental-sg-production.  When there is no driver, sg production starts from scratch.

> is it 'reloading' the source generator (if the sg generator changes)

*not yet*.  But that is hte plan in the near future.  more stuff has to happen first.  happy to explain in person.